sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala

```diff
@@ -150,37 +150,48 @@ class TopicImpl(private val probe: TestProbe, codec: MessageCodec) extends Topic
     anyFromMessage(msg.getMessage)
   }
 
-  override def expectOneTyped[T <: GeneratedMessageV3](clazz: Class[T]): TestKitMessage[T] =
+  override def expectOneTyped[T](clazz: Class[T]): TestKitMessage[T] =
     expectOneTyped(clazz, DefaultTimeout)
 
-  override def expectOneTyped[T <: GeneratedMessageV3](clazz: Class[T], timeout: time.Duration): TestKitMessage[T] = {
+  override def expectOneTyped[T](clazz: Class[T], timeout: time.Duration): TestKitMessage[T] = {
     val msg = probe.expectMsgType[EmitSingleCommand]
     val metadata = new MetadataImpl(msg.getMessage.getMetadata.entries)
-    val decodedMsg = codec.decodeMessage(ScalaPbAny(typeUrlFor(metadata), msg.getMessage.payload))
+    val scalaPb = ScalaPbAny(typeUrlFor(metadata), msg.getMessage.payload)
+
+    val decodedMsg = if (typeUrlFor(metadata).startsWith(JsonSupport.KALIX_JSON)) {
+      val objectMapper = JsonSupport.getObjectMapper
+      objectMapper.readerFor(clazz).readValue(msg.getMessage.payload.toByteArray)
+    } else {
+      codec.decodeMessage(scalaPb)
+    }
 
     val concreteType = TestKitMessageImpl.expectType(decodedMsg, clazz)
     TestKitMessageImpl(concreteType, metadata)
   }
 
   private def anyFromMessage(m: kalix.testkit.protocol.eventing_test_backend.Message): TestKitMessage[_] = {
     val metadata = new MetadataImpl(m.metadata.getOrElse(Metadata.defaultInstance).entries)
-    val anyMsg = codec.decodeMessage(ScalaPbAny(typeUrlFor(metadata), m.payload))
-    TestKitMessageImpl(anyMsg, metadata).asInstanceOf[TestKitMessage[_]]
+    val anyMsg = if (typeUrlFor(metadata).startsWith(JsonSupport.KALIX_JSON)) {
+      m.payload.toStringUtf8
+    } else {
+      codec.decodeMessage(ScalaPbAny(typeUrlFor(metadata), m.payload))
+    }
+    TestKitMessageImpl(anyMsg, metadata)
   }
 
   private def typeUrlFor(metadata: MetadataImpl): String = {
-    metadata
-      .get("ce-type")
-      .orElseGet { () =>
-        val contentType = metadata.get("Content-Type").asScala
-        contentType match {
-          case Some("text/plain; charset=utf-8") => "type.kalix.io/string"
-          case Some("application/octet-stream")  => "type.kalix.io/bytes"
-          case unknown =>
-            log.warn(s"Could not extract typeUrl from $unknown")
-            ""
-        }
-      }
+    val ceType = metadata.get("ce-type").asScala
+    val contentType = metadata.get("Content-Type").asScala
+
+    (ceType, contentType) match {
+      case (_, Some("text/plain; charset=utf-8")) => "type.kalix.io/string"
+      case (_, Some("application/octet-stream"))  => "type.kalix.io/bytes"
+      case (Some(t), Some("application/json"))    => s"json.kalix.io/$t"
+      case (Some(t), _)                           => s"type.googleapis.com/$t"
+      case (t, ct) =>
+        log.warn(s"Could not extract typeUrl from ce-type=$t content-type=$ct")
+        ""
```

Comment:
we are probably still missing some edge cases here I think.. I will need to double check and cover with some tests.